### ISSUE TITLE
Snowflake: Casting to structured array results in "unparseable section" error

### ISSUE THREAD

**Issue by karenderer** [sqlfluff/sqlfluff#6236]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Casting to a variant to a structured array is valid in [Snowflake](https://docs.snowflake.com/en/sql-reference/data-types-structured#casting-semi-structured-arrays-and-variants-to-structured-arrays) but is not currently supported in sqlfluff.

### Expected Behaviour

The expected behavior is that casting to a structured array would be parseable.

### Observed Behaviour

When casting to a structured array, sqlfluff returns errors like the following

 > Found unparseable section: '(integer)'

### How to reproduce

```sql
select
    blob:content::array(integer) as field1, -- unparseable
    cast(blob:content as array(integer)) as field2, -- parseable
from foo
```

### Dialect

```
dialect = snowflake
```

### Version

```
$ sqlfluff --version
sqlfluff, version 3.2.0
```

### Configuration

```
[sqlfluff]
dialect = snowflake
templater = jinja
sql_file_exts = .sql
rules =
    capitalisation.keywords,
    capitalisation.literals,
    capitalisation.functions,
    capitalisation.types,
    capitalisation.identifiers
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
